### DISCUSSION TITLE
Serial Bridge and Scripting: Handling long serial data string

### DISCUSSION THREAD

**Discussion by jkwim** [arendst/Tasmota#16497] [Q&A]
I am trying to use an ESP8266 to send/receive commands/responses to/from a solar inverter.
Have setup the communication via a MAX3232 base RS232 to TTL converter + Level shifter to 3.3V.

Currently using Tx/Rx pins on ESP.

I send command in the format:

`Serialsend5 5150494753B7A90D`

I receive a long string (HEX encoded ASCII string)

`RSL: RESULT = {"SerialReceived":"283233392E372034392E38203233392E372034392E382030363030203035343220303130203433302035342E3130203030302031303020303033362030302E30203033362E382030302E30302030303030302030303031303130312030302030302030303030302031313022340D"}`

Actually my preferred command would be 

`Serialsend3 \x51\x50\x49\x47\x53\xB7\xA9\x0D`

so that I could receive the ASCII string in the RESULT. 

However I suspect that that receive buffer conversion cannot handle that and gives me just a "/r" as a response. (It could be due to a CRC & \x0D appended at the end. Only the \x0D gets printed)

So comming back to Serialsend5 I wanted to try a script to read desired number of characters from the long string, convert them to ASCII and use the value in a MQTT message (+ web page).

I want to understand how to extract substrings directlly like following (this code does not work):

```
>D
p1=""
>E
;sb(svar p n) substring at position p (if p<0 counts from end) and length n
p1=sb(SerialReceived 0 8)
print First Par: %p1%
```

Since the received string is too long to be stored in a variable, I would like to get chunks of it and use them.

What is the correct syntax to be used to extract a substring from the receieved string?

**Comment by gemu2015** [comment]
check this discussion. there are example scripts how to implement solar inverters with arbitrary protocols 
https://github.com/arendst/Tasmota/discussions/15696
to decode longer strings you have to increase max script string size by e.g. #define SCRIPT_MAXSSIZE 128

**Reply by jkwim** [reply]
Great and Thank you!
Looks like a good starting point for my problem.
Will tinker your example and see.

**Reply by jkwim** [reply]
> check this discussion. there are example scripts how to implement solar inverters with arbitrary protocols #15696 to decode longer strings you have to increase max script string size by e.g. #define SCRIPT_MAXSSIZE 128

I have a question. The actual data string that is being received is an ASCII delimited string like:

`(239.1 50.0 239.1 50.0 0431 0297 007 449 54.10 000 100 0043 01.0.00 00000 10010111 00 00 00317 1109 <CRC> \r`

Since the parameter lengh in bytes is different for each parameter, it is not easy to get to the delimiter which is acctually \0x20 <space>

Is there a way I can read this in to a string variable directly so that I can extract using space as a delimiter?

Alternatively I tried something like this to read but does not work:
`res=sr(rstr \x20)`

How do you specify the space character as a delimiter anyway even in the following command?
`st(rstr , 3)`

**Reply by gemu2015** [reply]
to read a serial string until space is:
res=sr(rstr 0x20)

st() did not support invisible chars.
i added this already but not pr yet: 
st(rstr ' ' 3)
[xdrv_10_scripter.ino.zip](https://github.com/arendst/Tasmota/files/9553652/xdrv_10_scripter.ino.zip)

**Reply by jkwim** [reply]
I did progress further.

However as soon as I add the last nested for next loop (which is shown remerked) the script gets confused.

```
>S
; check if serial port received bytes
res=sa()
;print %res% bytes in buffer
if res>0 {
 ; read result into array
 ;res=sr(rstr 0x20)
 ;print rstr=%rstr%
 res=sra(rec_a);
 print %res% bytes read

 ; now analyze array
 if rec_a[1]==0x28 {
  print valid receive
  ; now analyze array for results
  acnt=0
  for cnt 1 res 1
   tvar=rec_a[cnt]
   print %cnt% = %tvar% acnt=%acnt%
   val_a[acnt]=tvar
   if tvar==0x20 {
    print space character tvar=%tvar% detected at acnt=%acnt% cnt=%cnt%
    ;for cnt2 0 acnt 1
     rstr=val_a[cnt-1]
     ;print counter2:%cnt2%
     print counter1:%cnt% counter2:%cnt2% string:%rstr%
    ;next
    acnt=0
    rstr=""
   }
   acnt+=1
   next
 }
}
```

**Reply by gemu2015** [reply]
loop nesting is  not supported!
you may call a subroutine which may then contain another loop

**Reply by gemu2015** [reply]
i just added nested loops, because several users requested it. ( up to 3 levels)
[xdrv_10_scripter.ino.zip](https://github.com/arendst/Tasmota/files/9556328/xdrv_10_scripter.ino.zip)

**Reply by jkwim** [reply]
> i just added nested loops, because several users requested it. ( up to 3 levels) [xdrv_10_scripter.ino.zip](https://github.com/arendst/Tasmota/files/9556328/xdrv_10_scripter.ino.zip)

Thank you.

I guess I want to try a soubroutine first as you had suggested before heading for another compiling/flashing. Then I can get the hang of subroutines as well.

**Reply by jkwim** [reply]
> > i just added nested loops, because several users requested it. ( up to 3 levels) [xdrv_10_scripter.ino.zip](https://github.com/arendst/Tasmota/files/9556328/xdrv_10_scripter.ino.zip)
> 
> Thank you.
> 
> I guess I want to try a soubroutine first as you had suggested before heading for another compiling/flashing. Then I can get the hang of subroutines as well.

I am stuck again.

I can detect and group the bytes for each ASCII string.

```
07:57:15.805 acnt=1.00 rec_a[13.00]=50.00 val_a[1.00]=50.00
07:57:15.808 acnt=2.00 rec_a[14.00]=51.00 val_a[2.00]=51.00
07:57:15.811 acnt=3.00 rec_a[15.00]=54.00 val_a[3.00]=54.00
07:57:15.814 acnt=4.00 rec_a[16.00]=46.00 val_a[4.00]=46.00
07:57:15.817 acnt=5.00 rec_a[17.00]=48.00 val_a[5.00]=48.00
07:57:15.820 acnt=6.00 rec_a[18.00]=32.00 val_a[6.00]=32.00
07:57:15.822 space character tvar=32.00 detected at cnt=18.00 acnt=6.00

07:57:15.946 acnt=1.00 rec_a[19.00]=52.00 val_a[1.00]=52.00
07:57:15.948 acnt=2.00 rec_a[20.00]=57.00 val_a[2.00]=57.00
07:57:15.951 acnt=3.00 rec_a[21.00]=46.00 val_a[3.00]=46.00
07:57:15.954 acnt=4.00 rec_a[22.00]=57.00 val_a[4.00]=57.00
07:57:15.957 acnt=5.00 rec_a[23.00]=32.00 val_a[5.00]=32.00
07:57:15.960 space character tvar=32.00 detected at cnt=23.00 acnt=5.00
```

`val_a` contains the grouped bytes.

Now I need to concatenate these bytes after converting each byte in to an ASCII character.

Looked in your code for all the `if (!strncmp(lp, ...` lines but could not figure out whether there is a function to get a HEX or Decimal value convertered in to an ASCII Character. 

And also a way to concatenate strings. 

Perhaps if new function can be written to 

Pick a substring using a delimiter and group the bytes and convert in to ASCII string on one go.

You already have a function to group 4 bytes and convert the in to float

af(array index) = converts 4 bytes of an array at index index to float number 

If a similar function can be created to just group specified number of bytes and convert in to an ASCII string then a generalized capability will be there to handle this kind of protocol (there are many solar inverters follow this protocol format).

Thanks for your support so far which has been great!

**Reply by jkwim** [reply]
One more thing. 
The last 3 bytes of the response need to be ignored. 2 x bytes for CRC16 and a carriage return (0x0D).

**Reply by jkwim** [reply]
Instead of just forwarding a protocol document I will simply explain the protocol structure.

1. Essentially they are ASCII commands like QPIGS, QPIRI, QMOD etc.

3. A XMODEM CRC(16) needs to be added for above ASCII string

4. Terminating character will be 0x0D

5. Since the majority of commands are fixed strings like in 1. above, the CRC can be manually calculated and appended to the byte stream. So will be the terminating characters.

6. For example I have shown below how the QPIGS command is constructed:
```
trx_a[1]=0x51    Q
trx_a[2]=0x50    P
trx_a[3]=0x49    I
trx_a[4]=0x47   G
trx_a[5]=0x53   S
trx_a[6]=0xB7   CRC16 byte 1
trx_a[7]=0xA9   CRC16 byte 2
trx_a[8]=0x0D   termination character
```

7. Resulting response will contain an ASCII string like this:
`(230.0 23.9 230.0 50.0 23.9 5500 5500 48.0 46.0 48.0 56.8 54.0 2 10 010 0 0 2 1 01 0 0 54.0 0 1ä`
  
   The starting character of the response is` '('` which is 0x28.

8. To extract the first value, we need to group the next 5 bytes after converting each byte in to ASCII chacater representation

9. Then comes a space character

10. Next 4 bytes represent the second value

11. As you can see the length of each value string is not fixed. The grouping can be acheived using the space character as the delimiter

12. In the response also there wll be a CRC16 and the terminating character. If you attempt to decode the entire string in to ASCII the CRC bytes may produce non printable characters. I think this is where the problem is when trying the Tasmota's `serialsend3` command. Tasmota tried to convert CRC16 bytes also in to ASCII.

13. If we tackle the bytes from the script, we can skip the last 3 bytes to avoid this issue.

Hope that this summerises what we are trying to get at.

**Reply by gemu2015** [reply]
since your meter response is pure asci except the checksum you should not use an array for receiving
set the max string size as proposed above to e.g. 128 and force this size by
\>D 128

you must assure that after you have requested the data from the meter you have passed some time for the string to arrive if the meter responses within a second, simply request the meter data in the end of the >S section

then in >S simply separate the received string into tokens by

res=sa()
if res>0 {
rstr=sr()
for cnt 1 22 1
str=st(rstr ' ' cnt)
if sl(str)==0 {
break
}
print %str%
next
}

or read each chunk to space.

res=sa()
if res>0 {
for cnt 1 22 1
str=sr(0x20)
if sl(str)==0 {
break
}
print %str%
next
}

btw to concatenate strings simply add them  
str="how"+" are"+" you ?"

to convert a number to a char was not yet implemented 
i just added the to char cmd: str=tc(num)

**Comment by jkwim** [comment]
OK here is the progress:
```
>D 128
; max string size 128

res=0
; transmit array
m:trx_a=0 30


cnt=0
cnt2=0
tvar=0
rstr=""
res2=0
str=""

>B
; preset cmd 51 50 49 47 53 B7 A9 0D QPIGS
trx_a[1]=0x51
trx_a[2]=0x50
trx_a[3]=0x49
trx_a[4]=0x47
trx_a[5]=0x53
trx_a[6]=0xB7
trx_a[7]=0xA9
trx_a[8]=0x0D
; open serial port, replace pin numbers with your actual wiring recpin, trxpin and baudrate
; tx gpio0 rx gpio3
res=so(3 1 2400 8N1 1024)

>S
; check if serial port received bytes
res=sa()
print %res% bytes in buffer
if res>0 {
 ; read result
 print reading in
 rstr=sr()

 ; remove first and last 3 characters
 rstr=sb(rstr 1 sl(rstr)-4)
 print rstr=%rstr%

 for cnt 1 22 1
  print processing chunks
  print cnt=%cnt%
  ;st(svar c n)  retrieve the nth element of svar delimited by c
  ;str=st(rstr ' ' cnt)
  tvar=sl(str)
  if sl(str)==0 {
   break
  }
  print length=%tvar% str=%str%
 next
}


>T
; send next cmd array with no checksum
print sending command
swa(trx_a 8)
```

```
20:02:59.134 110.00 bytes in buffer
20:02:59.135 reading in
20:02:59.138 rstr=235.0 50.0 235.0 50.0 0633 0600 011 431 54.10 000 100 0037 00.0 036.3 00.00 00000 00010101 00 00 00000 110
20:02:59.139 processing chunks
20:02:59.141 cnt=0.00                     <<<<<<<<< WHY IS THIS ZERO?
20:03:00.369 0.00 bytes in buffer
20:03:01.134 0.00 bytes in buffer
20:03:02.135 0.00 bytes in buffer
```

1. I just dropped in your latest .io file in to the refreshed archive and compiled. Hope that there were no additional steps required

2. The for next loop does not seem to work. The `cnt` value does not get set to `1` in the loop, but remains at initial value of `0`

**Reply by gemu2015** [reply]
i copied your loop and it works for me, but the break cmd seems to be broken.

use this until i fix the break:

for cnt 1 22 1
  print processing chunks
  print cnt=%cnt%
  str=st(rstr ' ' cnt)
  tvar=sl(str)
  if tvar>0 {
   print length=%tvar% str=%str%
  }
 next

cnt should never be zero though ???

**Reply by jkwim** [reply]
> i copied your loop and it works for me, but the break cmd seems to be broken.

It appears that your latest .io file's for next loop is sensitive to indentation. When I removed indentation it worked.

```
for cnt 1 22 1
  str=st(rstr ' ' cnt)
  tvar=sl(str)
  if tvar>0 {
   print cnt=%cnt% length=%tvar% str=%str%
  }
next
```

If I recall correctly the old version did not have this issue. Perhaps something that got introduced when allowing nested loops?

**Comment by jkwim** [comment]
A quick hack to get something working until we sort out the for next loop issue.

```
>W
Grid Voltage{m}%st(rstr ' ' 1)% V
Grid Frequency{m}%st(rstr ' ' 2)% Hz
AC Output Voltage{m}%st(rstr ' ' 3)% V
AC Output Frequency{m}%st(rstr ' ' 4)% Hz
AC Output Apparent Power{m}%st(rstr ' ' 5)% VA
AC Output Active Power{m}%st(rstr ' ' 6)% W
Output Load{m}%st(rstr ' ' 7)% \%
BUS Voltage{m}%st(rstr ' ' 8)% V
Battery Voltage{m}%st(rstr ' ' 9)% V
Battery Charging Current{m}%st(rstr ' ' 10)% A
Battery Capacity{m}%st(rstr ' ' 11)% \%
Inverter Heat Sink Temperature{m}%st(rstr ' ' 12)% ℃
PV Input Current for Battery{m}%st(rstr ' ' 13)% A
PV Input Voltage 1{m}%st(rstr ' ' 14)% V
Battery Voltage from SCC{m}%st(rstr ' ' 15)% V
Battery Discharge Current{m}%st(rstr ' ' 16)% A
```
![image](https://user-images.githubusercontent.com/84637540/190217584-2c4bf804-4ab4-49b7-a3c3-682429e6eed5.png)

**Reply by jkwim** [reply]
In order to capture responses from multiple commands and construct a single MQTT message I am thinking of using following logic

```
>D
cmd[0]
cmd[1]
cmd[2]
Outputarray


counter=0
timer


>S

counter is somevalue
At certain timer expiry:
  send cmd[counter]
delay
Read in the response
  analyse for cmd[counter]
  push values in to output array
increment counter
If counter==2{
  Now array will contain valus obtained from all 3 commands
  Process them eg: send out MQTT, update web page
  reset counter to 0
  clear the output array
}
```

Do you think that there is a better & easy way?

**Reply by jkwim** [reply]
Also, can you explain how string arrays are defined.
Following is not very clear to me from documentation:

is(num "string1|string2|....|stringn|") = defines a string array optionally preset with immediate strings separated by '|' (this immediate string may be up to 255 chars long) num = 0 read only string array, num > 0 number of elements in read write string array
is[index] = gets string index from string array, if read-write also write string of index
is1(..), is2(...) string array see above
is1[x], is2[x] string array see above

**Reply by gemu2015** [reply]
string arrays can be used in 2 ways
1. as providing string constants with efficient storage. (num=0, read only)
2. as real arrays of strings (num > 0, read write) this will consume num * SCRIPT_MAXSSIZE memory

in your case i would get rid of strings as your variables are actually numbers.
i would convert them to an array directly after receiving

for cnt 1 22 1
  str=st(rstr ' ' cnt)
  tvar=sl(str)
  if tvar>0 {
   array[cnt]=str
   print cnt=%cnt% length=%tvar% str=%str%
  }
next

**Reply by jkwim** [reply]
so in the documentation `'is'` is an array name rather than a function right?

then I need to be cautious about :
`
tvar=sl(str)
`
as this becomes 0 even for a valid reporting of value 000 by the inverter.

the issue is apparent in above code.

I will give a try for multipe commands. There two more commands that I need to send and consolidate the values.

Some of the values are really binary flags.

Are there any commands for converting in to binary digits and extracting each bit? ie. like biniary shifting

**Reply by gemu2015** [reply]
just a remark:
> then I need to be cautious about :
> tvar=sl(str)
> as this becomes 0 even for a valid reporting of value 000 by the inverter.

the string length of 000 will be 3 and not zero, so no issue

any binary operator is supported, see the docs

**Reply by jkwim** [reply]
I am back to scripting after setting up the end to end solution with Grafana.

At the moment the single command script works fine. I am generating a single MQTT message as following:

```
; MQTT export
>J
,"SAKO":{
"GridV":"%st(rstr ' ' 1)%",
"ActPw":"%st(rstr ' ' 6)%",
"OutV":"%st(rstr ' ' 3)%",
"BatV":"%st(rstr ' ' 9)%",
"BatVSCC":"%st(rstr ' ' 15)%",
"BatCap":"%st(rstr ' ' 11)%",
"BatChargCur":"%st(rstr ' ' 10)%",
"BatDischCur":"%st(rstr ' ' 16)%",
"PVPower":"%st(rstr ' ' 20)%",
"PVVoltage":"%st(rstr ' ' 14)%",
"PVCurrent":"%st(rstr ' ' 13)%",
"BUSV":"%st(rstr ' ' 8)%",
"InverterT":"%st(rstr ' ' 12)%"
}

```

I now need to be able to generate additional MQTT messages under certain conditions.

How do I construct the MQTT messages conditionally.

I figured out that the above contruction just appends everything to a single JSON string.

**Reply by gemu2015** [reply]
you may simply use the tasmota publish cmd:

if xxx {
=>publish ....
}

**Reply by jkwim** [reply]
OK. Got that partially working. 

How do you get the timestamp included?

`=>publish stat/Tasmota/INVERTER {"Time":"%timestamp%","ActPw":"%tvar%"};`

Rules fill fill in %timestamp% but through script I am getting:
`INVERTER = {"Time":"36.00","ActPw":"2674.00"};`

Need to be something like :
`
{"Time":"2022-09-24T00:37:25", ...
`

`

**Reply by gemu2015** [reply]
It is tstamp not timestamp, see scripting docs

**Reply by jkwim** [reply]
> It is tstamp not timestamp, see scripting docs

Thanks :-) 

Got it working!

**Comment by jkwim** [comment]
Hi,
Back here after few days. Haven't had much time to play with this.

Quick question:
Does the scripting interface allow opening of two serial ports?

I do not see any handler being setup when opening so I presume the answer would be no.

```
; open serial port, replace pin numbers with your actual wiring recpin, trxpin and baudrate
; tx gpio1 rx gpio3
res=so(3 1 2400 8N1 1024)



; check if serial port received bytes
res=sa()

```

**Reply by gemu2015** [reply]
yes currently only 1 serial port open at a time.

**Comment by Yarda81** [comment]
> I am trying to use an ESP8266 to send/receive commands/responses to/from a solar inverter. Have setup the communication via a MAX3232 base RS232 to TTL converter + Level shifter to 3.3V.
> 
> Currently using Tx/Rx pins on ESP.
> 
> I send command in the format:
> 
> `Serialsend5 5150494753B7A90D`
> 
> I receive a long string (HEX encoded ASCII string)
> 
> `RSL: RESULT = {"SerialReceived":"283233392E372034392E38203233392E372034392E382030363030203035343220303130203433302035342E3130203030302031303020303033362030302E30203033362E382030302E30302030303030302030303031303130312030302030302030303030302031313022340D"}`
> 
> Actually my preferred command would be
> 
> `Serialsend3 \x51\x50\x49\x47\x53\xB7\xA9\x0D`
> 
> so that I could receive the ASCII string in the RESULT.
> 
> However I suspect that that receive buffer conversion cannot handle that and gives me just a "/r" as a response. (It could be due to a CRC & \x0D appended at the end. Only the \x0D gets printed)
> 
> So comming back to Serialsend5 I wanted to try a script to read desired number of characters from the long string, convert them to ASCII and use the value in a MQTT message (+ web page).
> 
> I want to understand how to extract substrings directlly like following (this code does not work):
> 
> ```
> >D
> p1=""
> >E
> ;sb(svar p n) substring at position p (if p<0 counts from end) and length n
> p1=sb(SerialReceived 0 8)
> print First Par: %p1%
> ```
> 
> Since the received string is too long to be stored in a variable, I would like to get chunks of it and use them.
> 
> What is the correct syntax to be used to extract a substring from the receieved string?

Hello i am interesting connecting RS232 on my inverter to tasmota for reading by loxone system, can you take more informations about connected pins, scheme and so on, please?

**Reply by jkwim** [reply]
> > I am trying to use an ESP8266 to send/receive commands/responses to/from a solar inverter. Have setup the communication via a MAX3232 base RS232 to TTL converter + Level shifter to 3.3V.
> > Currently using Tx/Rx pins on ESP.
> > I send command in the format:
> > `Serialsend5 5150494753B7A90D`
> > I receive a long string (HEX encoded ASCII string)
> > `RSL: RESULT = {"SerialReceived":"283233392E372034392E38203233392E372034392E382030363030203035343220303130203433302035342E3130203030302031303020303033362030302E30203033362E382030302E30302030303030302030303031303130312030302030302030303030302031313022340D"}`
> > Actually my preferred command would be
> > `Serialsend3 \x51\x50\x49\x47\x53\xB7\xA9\x0D`
> > so that I could receive the ASCII string in the RESULT.
> > However I suspect that that receive buffer conversion cannot handle that and gives me just a "/r" as a response. (It could be due to a CRC & \x0D appended at the end. Only the \x0D gets printed)
> > So comming back to Serialsend5 I wanted to try a script to read desired number of characters from the long string, convert them to ASCII and use the value in a MQTT message (+ web page).
> > I want to understand how to extract substrings directlly like following (this code does not work):
> > ```
> > >D
> > p1=""
> > >E
> > ;sb(svar p n) substring at position p (if p<0 counts from end) and length n
> > p1=sb(SerialReceived 0 8)
> > print First Par: %p1%
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Since the received string is too long to be stored in a variable, I would like to get chunks of it and use them.
> > What is the correct syntax to be used to extract a substring from the receieved string?
> 
> Hello i am interesting connecting RS232 on my inverter to tasmota for reading by loxone system, can you take more informations about connected pins, scheme and so on, please?

Without knowing your inverter details I cannot help much.

Basically you would need to connect:

Tx
Rx
GND

pins from the inverter to a 5V to 3.3V Logic Level Shifter and then connect the 3.3V Rx/Tx and GND to ESP's Tx/Rx/GND.

Some inverters may give out 12V from the RS-232 port using an otherwise unused pin.

You need to connnect this 12V to a 12V-5V DCtoDC Buck Converter and obtain.

If you use a ESPBoard like Node MCU/Wemos D1 Mini then you can feed in 5V and then take 3.3V out from the board to wire the 3.3V side of the logic level converter. If you intend to use a module like ESP12E/F then you need to use an addional 5V to 3.3V regulator (ie. ASM1117) also.

Roughly this is how I have built my Hardware setup.

You will need to adapt to your inverters sonfiguration.
